File: java/norito_java/src/main/java/org/hyperledger/iroha/norito/NoritoEncoder.java
// Copyright 2024 Hyperledger Iroha Contributors
// SPDX-License-Identifier: Apache-2.0

package org.hyperledger.iroha.norito;

import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Objects;

/** Low-level helper for writing Norito payloads. */
public final class NoritoEncoder {
  private final int flags;
  private final ByteArrayOutputStream buffer;

  public NoritoEncoder(int flags) {
    this.flags = flags;
    this.buffer = new ByteArrayOutputStream();
  }

  public int flags() {
    return flags;
  }

  public void writeByte(int value) {
    buffer.write(value & 0xFF);
  }

  public void writeBytes(byte[] data) {
    buffer.writeBytes(data);
  }

  public void writeUInt(long value, int bits) {
    int bytes = bits / 8;
    ByteBuffer bb = ByteBuffer.allocate(bytes).order(ByteOrder.LITTLE_ENDIAN);
    if (bits == 64) {
      bb.putLong(value);
    } else if (bits == 32) {
      bb.putInt((int) value);
    } else if (bits == 16) {
      bb.putShort((short) value);
    } else if (bits == 8) {
      bb.put((byte) value);
    } else {
      throw new IllegalArgumentException("Unsupported integer size: " + bits);
    }
    writeBytes(bb.array());
  }

  public void writeInt(long value, int bits) {
    writeUInt(value, bits); // little-endian same for signed writes
  }

  public void writeLength(long value, boolean compact) {
    if (compact) {
      writeBytes(Varint.encode(value));
    } else {
      writeUInt(value, 64);
    }
  }

  public NoritoEncoder childEncoder() {
    return new NoritoEncoder(flags);
  }

  public byte[] toByteArray() {
    return buffer.toByteArray();
  }

  public void append(byte[] data) {
    writeBytes(Objects.requireNonNull(data));
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/norito/NoritoEncoder.kt
// Copyright 2024 Hyperledger Iroha Contributors
// SPDX-License-Identifier: Apache-2.0

package org.hyperledger.iroha.sdk.norito

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

/** Low-level helper for writing Norito payloads. */
class NoritoEncoder(@JvmField val flags: Int) {

    private val buffer = ByteArrayOutputStream()

    fun writeByte(value: Int) {
        buffer.write(value and 0xFF)
    }

    fun writeBytes(data: ByteArray) {
        buffer.write(data, 0, data.size)
    }

    fun writeUInt(value: Long, bits: Int) {
        val bytes = bits / 8
        val bb = ByteBuffer.allocate(bytes).order(ByteOrder.LITTLE_ENDIAN)
        when (bits) {
            64 -> bb.putLong(value)
            32 -> bb.putInt(value.toInt())
            16 -> bb.putShort(value.toShort())
            8 -> bb.put(value.toByte())
            else -> throw IllegalArgumentException("Unsupported integer size: $bits")
        }
        writeBytes(bb.array())
    }

    fun writeInt(value: Long, bits: Int) {
        writeUInt(value, bits)
    }

    fun writeLength(value: Long, compact: Boolean) {
        if (compact) {
            writeBytes(Varint.encode(value))
        } else {
            writeUInt(value, 64)
        }
    }

    fun childEncoder(): NoritoEncoder = NoritoEncoder(flags)

    fun toByteArray(): ByteArray = buffer.toByteArray()

    fun append(data: ByteArray) {
        writeBytes(data)
    }
}
